src/order/disjointed.lean

```diff
@@ -8,83 +8,139 @@ import order.partial_sups
+* `disjoint_disjointed`: The elements of `disjointed f` are pairwise disjoint.
 -/
 
 variables {α β : Type*}
 
-namespace set
+section generalized_boolean_algebra
+variables [generalized_boolean_algebra α]
+
+/-- If `f : ℕ → α` is a sequence of elements, then `disjointed f` is the sequence formed by
+subtracting each element from the nexts. This is the unique disjoint sequence whose partial sups
+are the same as the original sequence. -/
+def disjointed (f : ℕ → α) : ℕ → α
+| 0       := f 0
+| (n + 1) := f (n + 1) \ (partial_sups f n)
 
-/-- If `f : ℕ → set α` is a sequence of sets, then `disjointed f` is
-  the sequence formed with each set subtracted from the later ones
-  in the sequence, to form a disjoint sequence. -/
-def disjointed (f : ℕ → set α) (n : ℕ) : set α := f n ∩ (⋂ i < n, (f i)ᶜ)
+lemma disjointed_zero (f : ℕ → α) : disjointed f 0 = f 0 := rfl
 
-variables {f : ℕ → set α} {n : ℕ}
+lemma disjointed_succ (f : ℕ → α) (n : ℕ) :
+  disjointed f (n + 1) = f (n + 1) \ (partial_sups f n) :=
+rfl
 
-lemma disjoint_disjointed : pairwise (disjoint on disjointed f) :=
-λ i j h, begin
-  wlog h' : i ≤ j; [skip, {revert a, exact (this h.symm).symm}],
-  rintro a ⟨⟨h₁, _⟩, h₂, h₃⟩, simp at h₃,
-  exact h₃ _ (lt_of_le_of_ne h' h) h₁
+lemma disjointed_le_id : disjointed ≤ (id : (ℕ → α) → ℕ → α) :=
+begin
+  rintro f n,
+  cases n,
+  { refl },
+  { exact sdiff_le }
 end
 
--- a more useful version might be `∀ i j x, x ∈ disjointed f i → x ∈ disjointed f j → i = j`
-lemma disjoint_disjointed' :
-  ∀ i j, i ≠ j → (disjointed f i) ∩ (disjointed f j) = ∅ :=
-λ i j hij, disjoint_iff.1 $ disjoint_disjointed i j hij
+lemma disjointed_le (f : ℕ → α) : disjointed f ≤ f := disjointed_le_id f
 
-lemma disjointed_subset : disjointed f n ⊆ f n := inter_subset_left _ _
+lemma disjoint_disjointed (f : ℕ → α) : pairwise (disjoint on disjointed f) :=
+begin
+  rw symmetric.pairwise_on disjoint.symm, swap, apply_instance,
+  rintro m n h,
+  cases n,
+  { exact (h.not_le (nat.zero_le _)).elim },
+  exact disjoint_sdiff_self_right.mono_left ((disjointed_le f m).trans
+    (le_partial_sups_of_le f (nat.lt_add_one_iff.1 h))),
+end
 
-lemma Union_lt_succ : (⋃ i < nat.succ n, f i) = f n ∪ (⋃ i < n, f i) :=
-ext $ λ a, by simp [nat.lt_succ_iff_lt_or_eq, or_and_distrib_right, exists_or_distrib, or_comm]
+/-- An induction principle for `disjointed`. To define/prove something on `disjointed f n`, it's
+enough to define/prove it for `f n` and being able to extend through diffs. -/
+def disjointed_induct {f : ℕ → α} {p : α → Sort*} (hdiff : ∀ ⦃t i⦄, p t → p (t \ f i)) :
+  ∀ ⦃n⦄, p (f n) → p (disjointed f n)
+| 0       := id
+| (n + 1) := λ h,
+  begin
+    rw disjointed_succ,
+    suffices H : ∀ k, p (f (n + 1) \ partial_sups f k),
+    { exact H n },
+    rintro k,
+    induction k with k ih,
+    { rw partial_sups_zero,
+      exact hdiff h },
+    rw [partial_sups_succ, ←sdiff_sdiff_left],
+    exact hdiff ih,
+  end
```

Comment:
Since you've promoted this to `Sort*`, you're now responsible for providing simp lemmas for what this computes when in `Type`.

Reply:
Uh. I can easily prove `disjointed_rec_zero`, but `disjointed_rec_succ` involves an induction itself. I'm not sure I can say anything meaningful about it, except maybe using `nat.rec`?

Reply:
Let's just add the `rec_zero` version, and maybe a `/-! TODO -/` comment about `disjointed_rec_succ`.

Reply:
Note that it's actually not a recursion, but a case disjunction. Probably `rec` is the wrong name for it.